src/cascadia/TerminalSettingsModel/VsSetupConfiguration.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "VsSetupConfiguration.h"
+
+using namespace Microsoft::Terminal::Settings::Model;
+
+std::vector<VsSetupConfiguration::VsSetupInstance> VsSetupConfiguration::QueryInstances()
+{
+    std::vector<VsSetupConfiguration::VsSetupInstance> instances;
+
+    // SetupConfiguration is only registered if Visual Studio is installed
+    ComPtrSetupQuery pQuery{ wil::CoCreateInstanceNoThrow<SetupConfiguration, ISetupConfiguration2>() };
+    if (pQuery == nullptr)
+    {
+        return instances;
+    }
+
+    // Enumerate all valid instances of Visual Studio
+    wil::com_ptr<IEnumSetupInstances> e;
+    THROW_IF_FAILED(pQuery->EnumInstances(&e));
+
+    std::array<ComPtrSetupInstance, 1> rgpInstance;
+    auto result = e->Next(1, &rgpInstance[0], NULL);
+    while (S_OK == result)
+    {
+        // wrap the COM pointers in a friendly interface
+        instances.emplace_back(VsSetupConfiguration::VsSetupInstance{ pQuery, rgpInstance[0] });
+        result = e->Next(1, &rgpInstance[0], NULL);
+    }
+
+    THROW_IF_FAILED(result);
+
+    return instances;
+}
+
+/// <summary>
+/// Takes a relative path under a Visual Studio installation and returns the absolute path.
+/// </summary>
+std::wstring VsSetupConfiguration::ResolvePath(ComPtrSetupInstance pInst, std::wstring_view relativePath)
+{
+    wil::unique_bstr bstrAbsolutePath;
+    THROW_IF_FAILED(pInst->ResolvePath(relativePath.data(), &bstrAbsolutePath));
+    return bstrAbsolutePath.get();
+}
+
+/// <summary>
+/// Determines whether a Visual Studio installation version falls within a specified range.
+/// The range is specified as a string, ex: "[15.0.0.0,)", "[15.0.0.0, 16.7.0.0)
+/// </summary>
+bool VsSetupConfiguration::InstallationVersionInRange(ComPtrSetupQuery pQuery, ComPtrSetupInstance pInst, std::wstring_view range)
+{
+    auto helper = pQuery.query<ISetupHelper>();
+
+    // VS versions in a string format such as "16.3.0.0" can be easily compared
+    // by parsing them into 64-bit unsigned integers using the stable algorithm
+    // provided by ParseVersion and ParseVersionRange
+
+    unsigned long long minVersion{ 0 };
+    unsigned long long maxVersion{ 0 };
+    THROW_IF_FAILED(helper->ParseVersionRange(range.data(), &minVersion, &maxVersion));

```

Comment:
FWIW: In our specific case, the string views we are passing *are* null-terminated. That's a side effect of them being compiled-in; there's no general guarantee that a string_view is null-terminated. This will work fine, but we'll want to be careful if we ever allow people to specify their own version ranges!